File: ios/Wavepoint/Auth/AccountDeletionService.swift
import Foundation
import Supabase

protocol AccountDeleting: Sendable {
  func deleteAccount() async throws
}

protocol AccountDeletionHTTPTransport: Sendable {
  func send(_ request: URLRequest) async throws -> HTTPURLResponse
}

struct URLSessionAccountDeletionTransport: AccountDeletionHTTPTransport {
  let session: URLSession

  init(session: URLSession = .shared) {
    self.session = session
  }

  func send(_ request: URLRequest) async throws -> HTTPURLResponse {
    let (_, response) = try await session.data(for: request)
    guard let response = response as? HTTPURLResponse else {
      throw AccountDeletionError.invalidResponse
    }
    return response
  }
}

struct SupabaseAccountDeletionService: AccountDeleting {
  private let functionURL: URL
  private let publishableKey: String
  private let transport: any AccountDeletionHTTPTransport
  private let accessToken: @Sendable () async throws -> String

  init(
    functionURL: URL,
    publishableKey: String,
    transport: any AccountDeletionHTTPTransport = URLSessionAccountDeletionTransport(),
    accessToken: @escaping @Sendable () async throws -> String
  ) {
    self.functionURL = functionURL
    self.publishableKey = publishableKey
    self.transport = transport
    self.accessToken = accessToken
  }

  init(
    client: SupabaseClient,
    configuration: AppConfiguration,
    transport: any AccountDeletionHTTPTransport = URLSessionAccountDeletionTransport()
  ) {
    self.init(
      functionURL: configuration.supabaseURL
        .appending(path: "functions/v1/delete-account"),
      publishableKey: configuration.supabasePublishableKey,
      transport: transport,
      accessToken: { try await client.auth.session.accessToken }
    )
  }

  func deleteAccount() async throws {
    var request = URLRequest(url: functionURL)
    request.httpMethod = "DELETE"
    request.setValue("Bearer \(try await accessToken())", forHTTPHeaderField: "Authorization")
    request.setValue(publishableKey, forHTTPHeaderField: "apikey")

    let response = try await transport.send(request)
    switch response.statusCode {
    case 204:
      return
    case 401:
      throw AccountDeletionError.authorizationExpired
    default:
      throw AccountDeletionError.httpStatus(response.statusCode)
    }
  }
}

enum AccountDeletionError: LocalizedError, Equatable {
  case authorizationExpired
  case invalidResponse
  case httpStatus(Int)

  var errorDescription: String? {
    switch self {
    case .authorizationExpired:
      "Reconnect Spotify, then try deleting your account again."
    case .invalidResponse:
      "Wavepoint could not confirm account deletion. Your account was not deleted."
    case .httpStatus:
      "Please try again. Your account was not deleted."
    }
  }
}

File: android/app/src/main/java/ai/mapier/swipe/account/AccountDeletionService.kt
package ai.mapier.swipe.account

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request

interface AccountDeleting {
  suspend fun deleteAccount()
}

data class AccountDeletionRequest(
  val method: String,
  val url: String,
  val headers: Map<String, String>,
)

data class AccountDeletionResponse(val statusCode: Int)

interface AccountDeletionTransport {
  suspend fun send(request: AccountDeletionRequest): AccountDeletionResponse
}

class OkHttpAccountDeletionTransport(
  private val client: OkHttpClient = OkHttpClient(),
) : AccountDeletionTransport {
  override suspend fun send(request: AccountDeletionRequest): AccountDeletionResponse =
    withContext(Dispatchers.IO) {
      val builder = Request.Builder().url(request.url)
      request.headers.forEach(builder::header)
      if (request.method == "DELETE") builder.delete()
      client.newCall(builder.build()).execute().use { response ->
        AccountDeletionResponse(response.code)
      }
    }
}

enum class AccountDeletionErrorKind {
  AUTHORIZATION_EXPIRED,
  HTTP_STATUS,
}

class AccountDeletionException(
  val kind: AccountDeletionErrorKind,
  val statusCode: Int? = null,
) : Exception(
  when (kind) {
    AccountDeletionErrorKind.AUTHORIZATION_EXPIRED ->
      "Reconnect Spotify, then try deleting your account again."
    AccountDeletionErrorKind.HTTP_STATUS ->
      "Please try again. Your Wavepoint account was not deleted."
  },
)

class SupabaseAccountDeletionService(
  private val functionUrl: String,
  private val publishableKey: String,
  private val transport: AccountDeletionTransport = OkHttpAccountDeletionTransport(),
  private val accessToken: suspend () -> String,
) : AccountDeleting {
  override suspend fun deleteAccount() {
    val response = transport.send(
      AccountDeletionRequest(
        method = "DELETE",
        url = functionUrl,
        headers = mapOf(
          "Authorization" to "Bearer ${accessToken()}",
          "apikey" to publishableKey,
        ),
      ),
    )
    when (response.statusCode) {
      204 -> Unit
      401 -> throw AccountDeletionException(AccountDeletionErrorKind.AUTHORIZATION_EXPIRED)
      else -> throw AccountDeletionException(
        AccountDeletionErrorKind.HTTP_STATUS,
        statusCode = response.statusCode,
      )
    }
  }
}
